### ISSUE TITLE
FoundationDB pods doesn't start with TLS enabled

### ISSUE THREAD

**Issue by subbu165** [FoundationDB/fdb-kubernetes-operator#709]
With TLS disabled its working fine. but once we enable the TLS like below when creating the 'FoundationDBCluster' instance.
mainContainer:
  enableTls: true
sidecarContainer:
  enableTls: true

We found that after this, the configMap is populated empty like given below:
![image](https://user-images.githubusercontent.com/19837453/119380397-49b08980-bcde-11eb-8cba-4f32e4a872b7.png)

This is the example we followed for enabling TLS as sample
https://github.com/FoundationDB/fdb-kubernetes-operator/blob/master/config/samples/cluster_local_tls.yaml

As we see the operator logs, below is the only error we spot:
2021-05-24T07:53:22.827Z	ERROR	controller	Error in reconciliation	{"subReconciler": "controllers.GenerateInitialClusterFile", "namespace": "ABC", "cluster": "foundationdb-162184", "error": "missing one or more TLS env vars: FDB_TLS_CERTIFICATE_FILE, FDB_TLS_KEY_FILE or FDB_TLS_CA_FILE"}
github.com/go-logr/zapr.(*zapLogger).Error
	/go/pkg/mod/github.com/go-logr/zapr@v0.1.0/zapr.go:128
github.com/FoundationDB/fdb-kubernetes-operator/controllers.(*FoundationDBClusterReconciler).Reconcile
	/workspace/controllers/cluster_controller.go:152
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.4.0/pkg/internal/controller/controller.go:256
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.4.0/pkg/internal/controller/controller.go:232
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.4.0/pkg/internal/controller/controller.go:211
k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1
	/go/pkg/mod/k8s.io/apimachinery@v0.17.0/pkg/util/wait/wait.go:152
k8s.io/apimachinery/pkg/util/wait.JitterUntil
	/go/pkg/mod/k8s.io/apimachinery@v0.17.0/pkg/util/wait/wait.go:153
k8s.io/apimachinery/pkg/util/wait.Until
	/go/pkg/mod/k8s.io/apimachinery@v0.17.0/pkg/util/wait/wait.go:88

**Comment by subbu165** [comment]
This is the describe of whole FoundationDBCluster instance

apiVersion: apps.foundationdb.org/v1beta1
kind: FoundationDBCluster
metadata:
  selfLink: >-
    /apis/apps.foundationdb.org/v1beta1/namespaces/fdb/foundationdbclusters/foundationdb-1621596797112332
  resourceVersion: '3124562'
  name: foundationdb-1621596797112332
  uid: d0e297a7-ba8b-4893-89a2-fe1246241009
  creationTimestamp: '2021-05-21T11:34:04Z'
  generation: 1
  namespace: abc
  ownerReferences:
    - apiVersion: example.com/v1
      kind: SampleApp
      name: testinst
      uid: aac42537-6a20-4cab-8d54-6814fa452c3e
  labels:
    icpd-addon/status: '1621596797112332'
    app.kubernetes.io/part-of: foundationdb
    app.kubernetes.io/instance: '1621596797112332'
    app.kubernetes.io/version: 0.1.341-fdb-secret-9272-SNAPSHOT
    app.kubernetes.io/component: foundationdb
    icpdsupport/app: Sample-App
    icpdsupport/serviceInstanceId: '1621596797112332'
    app.kubernetes.io/name: foundationdb
    icpdsupport/addOnName: testinst
spec:
  faultDomain:
    key: foundationdb.org/none
  mainContainer:
    enableTls: true
  processCounts:
    cluster_controller: 1
    stateless: -1
  processes:
    general:
      customParameters:
        - knob_disable_posix_kernel_aio=1
      podTemplate:
        spec:
          initContainers:
            - image: >-
                example.io/foundationdb-kubernetes-sidecar
              name: foundationdb-kubernetes-init
              securityContext:
                allowPrivilegeEscalation: false
                capabilities:
                  drop:
                    - ALL
                privileged: false
                readOnlyRootFilesystem: false
          serviceAccountName: zen-norbac-sa
          imagePullSecrets:
            - name: abc-artifactory
          hostPID: false
          hostNetwork: false
          securityContext:
            runAsNonRoot: true
          containers:
            - env:
                - name: FDB_TLS_CERTIFICATE_FILE
                  value: /tmp/fdb-certs/tls.crt
                - name: FDB_TLS_CA_FILE
                  value: /tmp/fdb-certs/tls.crt
                - name: FDB_TLS_KEY_FILE
                  value: /tmp/fdb-certs/tls.key
              image: example.io/foundationdb
              name: foundationdb
              resources:
                limits:
                  cpu: 1000m
                  memory: 4Gi
                requests:
                  cpu: 500m
                  memory: 2Gi
              securityContext:
                allowPrivilegeEscalation: false
                capabilities:
                  drop:
                    - ALL
                privileged: false
                readOnlyRootFilesystem: false
              volumeMounts:
                - mountPath: /tmp/fdb-certs
                  name: fdb-kubernetes-operator-secrets-1621596797112332
            - env:
                - name: FDB_TLS_CERTIFICATE_FILE
                  value: /tmp/fdb-certs/tls.crt
                - name: FDB_TLS_CA_FILE
                  value: /tmp/fdb-certs/tls.crt
                - name: FDB_TLS_KEY_FILE
                  value: /tmp/fdb-certs/tls.key
              image: >-
                example.io/foundationdb-kubernetes-sidecar
              name: foundationdb-kubernetes-sidecar
              securityContext:
                allowPrivilegeEscalation: false
                capabilities:
                  drop:
                    - ALL
                privileged: false
                readOnlyRootFilesystem: false
              volumeMounts:
                - mountPath: /tmp/fdb-certs
                  name: fdb-kubernetes-operator-secrets-1621596797112332
          hostIPC: false
          volumes:
            - name: fdb-kubernetes-operator-secrets-1621596797112332
              secret:
                secretName: fdb-kubernetes-operator-secrets-1621596797112332
      volumeClaim:
        spec:
          resources:
            requests:
              storage: 16Gi
          storageClassName: portworx-shared-gp3
  sidecarContainer:
    enableTls: true
  version: 6.2.22
status:
  databaseConfiguration:
    log_routers: -1
    redundancy_mode: double
    remote_logs: -1
    storage_engine: ssd-2
    usable_regions: 1
  generations:
    needsConfigurationChange: 1
  health: {}
  missingProcesses:
    storage-2: 1621596869
    storage-3: 1621596869
    cluster_controller-1: 1621596869
    cluster_controller-2: 1621596870
    log-1: 1621596869
    log-2: 1621596869
    log-3: 1621596869
    log-4: 1621596869
    storage-1: 1621596869
  processCounts:
    cluster_controller: 2
    log: 4
    storage: 3
  requiredAddresses:
    tls: true
  runningVersion: 6.2.22

**Comment by subbu165** [comment]
ConfigMap content:

kind: ConfigMap
apiVersion: v1
metadata:
  selfLink: >-
    /api/v1/namespaces/abc/configmaps/foundationdb-1621620757762801-config
  resourceVersion: '3529195'
  name: foundationdb-1621620757762801-config
  uid: 643cd86e-8db1-42bb-a7fa-ffaeb8a8bf45
  creationTimestamp: '2021-05-21T18:13:55Z'
  namespace: abc
  ownerReferences:
    - apiVersion: apps.foundationdb.org/v1beta1
      kind: FoundationDBCluster
      name: foundationdb-1621620757762801
      uid: 90970d02-a4b3-4cb1-b500-43bde390105a
      controller: true
  labels:
    fdb-cluster-name: foundationdb-1621620757762801
data:
  cluster-file: ''
  fdbmonitor-conf-cluster_controller: ''
  fdbmonitor-conf-log: ''
  fdbmonitor-conf-storage: ''
  running-version: 6.2.22

**Comment by johscheuer** [comment]
Just to clarify you want to migrate a non-tls cluster to a tls cluster or just create a tls cluster? What version of the operator are you running?

**Comment by johscheuer** [comment]
Could you also verify that the generated Pods contain the env variables? That's what the controller complains about.

**Comment by subbu165** [comment]
Just create a tls cluster, No Migration. 
The sidecar pod version seems to be 6.2.22 as given in the above describe.
The operator Image is fdb-kubernetes-operator:0.23.1

**Comment by subbu165** [comment]
> Could you also verify that the generated Pods contain the env variables? That's what the controller complains about. And the ConfigMap is empty because the controller was not able to generate the initial cluster file (since the env variables `FDB_TLS_CERTIFICATE_FILE, FDB_TLS_KEY_FILE or FDB_TLS_CA_FILE` are missing).

This is what we done to confirm this, we logged into the side car pod and printed that env variables and they were present.

oc exec -it foundationdb-1621842727075674-cluster-controller-1 -c foundationdb-kubernetes-sidecar bash


![image](https://user-images.githubusercontent.com/19837453/119387455-9730f480-bce6-11eb-8725-d7811337ef2f.png)

**Comment by johscheuer** [comment]
Oh okay, so `0.23.1` is a while back you may want to update the operator version. So I guess you are missing these variables at the operator. The operator also needs these env variables to be able to communicate securely with the cluster.

**Comment by subbu165** [comment]
oh okay, what should be the right version that we should be using?

> Oh okay, so `0.23.1` is a while back you may want to update the operator version. So I guess you are missing these variables at the operator. The operator also needs these env variables to be able to communicate securely with the cluster.

**Comment by johscheuer** [comment]
The latest version (https://github.com/FoundationDB/fdb-kubernetes-operator/releases) would be `0.35.0` currently but the issue is actually not the version rather that you must add:

```yaml
- env:
  - name: FDB_TLS_CERTIFICATE_FILE
    value: /tmp/fdb-certs/tls.crt
  - name: FDB_TLS_CA_FILE
    value: /tmp/fdb-certs/tls.crt
  - name: FDB_TLS_KEY_FILE
    value: /tmp/fdb-certs/tls.key
```

and the secret mount to the operator deployment.

**Comment by johscheuer** [comment]
We need to actually document that 🤔

**Comment by subbu165** [comment]
Okay, we will try adding this env in the opeartor config and test it. Will update the thread here.

**Comment by subbu165** [comment]
@johscheuer we have included the env as below:
![image](https://user-images.githubusercontent.com/19837453/119396285-4c1cde80-bcf2-11eb-9d05-5269501b5b49.png)

It throws a python error as below:
TASK [Change FoundationDB manager state to "{{ state }}"] ******************************** 
The full traceback is:
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/ansible/executor/task_executor.py", line 592, in _execute
    self._task.post_validate(templar=templar)
  File "/usr/local/lib/python3.8/site-packages/ansible/playbook/task.py", line 299, in post_validate
    super(Task, self).post_validate(templar)
  File "/usr/local/lib/python3.8/site-packages/ansible/playbook/base.py", line 431, in post_validate
    value = templar.template(getattr(self, name))
  File "/usr/local/lib/python3.8/site-packages/ansible/template/__init__.py", line 657, in template
    d[k] = self.template(
  File "/usr/local/lib/python3.8/site-packages/ansible/template/__init__.py", line 611, in template
    result = self.do_template(
  File "/usr/local/lib/python3.8/site-packages/ansible/template/__init__.py", line 876, in do_template
    res = j2_concat(rf)
  File "<template>", line 13, in root
  File "/usr/local/lib/python3.8/site-packages/ansible/plugins/filter/core.py", line 210, in from_yaml
    return yaml.safe_load(data)
  File "/usr/local/lib64/python3.8/site-packages/yaml/__init__.py", line 162, in safe_load
    return load(stream, SafeLoader)
  File "/usr/local/lib64/python3.8/site-packages/yaml/__init__.py", line 114, in load
    return loader.get_single_data()
  File "/usr/local/lib64/python3.8/site-packages/yaml/constructor.py", line 49, in get_single_data
    node = self.get_single_node()
  File "/usr/local/lib64/python3.8/site-packages/yaml/composer.py", line 36, in get_single_node
    document = self.compose_document()
  File "/usr/local/lib64/python3.8/site-packages/yaml/composer.py", line 55, in compose_document
    node = self.compose_node(None, None)
  File "/usr/local/lib64/python3.8/site-packages/yaml/composer.py", line 84, in compose_node
    node = self.compose_mapping_node(anchor)
  File "/usr/local/lib64/python3.8/site-packages/yaml/composer.py", line 133, in compose_mapping_node
    item_value = self.compose_node(node, item_key)
  File "/usr/local/lib64/python3.8/site-packages/yaml/composer.py", line 84, in compose_node
    node = self.compose_mapping_node(anchor)
  File "/usr/local/lib64/python3.8/site-packages/yaml/composer.py", line 133, in compose_mapping_node
    item_value = self.compose_node(node, item_key)
  File "/usr/local/lib64/python3.8/site-packages/yaml/composer.py", line 84, in compose_node
    node = self.compose_mapping_node(anchor)
  File "/usr/local/lib64/python3.8/site-packages/yaml/composer.py", line 133, in compose_mapping_node
    item_value = self.compose_node(node, item_key)
  File "/usr/local/lib64/python3.8/site-packages/yaml/composer.py", line 84, in compose_node
    node = self.compose_mapping_node(anchor)
  File "/usr/local/lib64/python3.8/site-packages/yaml/composer.py", line 133, in compose_mapping_node
    item_value = self.compose_node(node, item_key)
  File "/usr/local/lib64/python3.8/site-packages/yaml/composer.py", line 82, in compose_node
    node = self.compose_sequence_node(anchor)
  File "/usr/local/lib64/python3.8/site-packages/yaml/composer.py", line 111, in compose_sequence_node
    node.value.append(self.compose_node(node, index))
  File "/usr/local/lib64/python3.8/site-packages/yaml/composer.py", line 84, in compose_node
    node = self.compose_mapping_node(anchor)
  File "/usr/local/lib64/python3.8/site-packages/yaml/composer.py", line 133, in compose_mapping_node
    item_value = self.compose_node(node, item_key)
  File "/usr/local/lib64/python3.8/site-packages/yaml/composer.py", line 82, in compose_node
    node = self.compose_sequence_node(anchor)
  File "/usr/local/lib64/python3.8/site-packages/yaml/composer.py", line 111, in compose_sequence_node
    node.value.append(self.compose_node(node, index))
  File "/usr/local/lib64/python3.8/site-packages/yaml/composer.py", line 84, in compose_node
    node = self.compose_mapping_node(anchor)
  File "/usr/local/lib64/python3.8/site-packages/yaml/composer.py", line 127, in compose_mapping_node
    while not self.check_event(MappingEndEvent):
  File "/usr/local/lib64/python3.8/site-packages/yaml/parser.py", line 98, in check_event
    self.current_event = self.state()
  File "/usr/local/lib64/python3.8/site-packages/yaml/parser.py", line 428, in parse_block_mapping_key
    if self.check_token(KeyToken):
  File "/usr/local/lib64/python3.8/site-packages/yaml/scanner.py", line 116, in check_token
    self.fetch_more_tokens()
  File "/usr/local/lib64/python3.8/site-packages/yaml/scanner.py", line 223, in fetch_more_tokens
    return self.fetch_value()
  File "/usr/local/lib64/python3.8/site-packages/yaml/scanner.py", line 577, in fetch_value
    raise ScannerError(None, None,
yaml.scanner.ScannerError: mapping values are not allowed here
  in "<unicode string>", line 79, column 22:
                    value: /tmp/fdb-certs/tls.crt
                         ^
fatal: [localhost]: FAILED! => {
    "changed": false
}

**Comment by johscheuer** [comment]
That's an Ansible/jinja error. From your screenshot it seems fine but there could be a whitespace (tabs vs. spaces) issue or something like that. 

This is the config for you development setup in this repo: https://github.com/FoundationDB/fdb-kubernetes-operator/blob/master/config/default/test_certs.yaml

**Comment by amankssm** [comment]
Now we are getting this issue. 
```
2021-05-24T20:16:35.651Z	ERROR	controller	Error in reconciliation	{"subReconciler": "controllers.GenerateInitialClusterFile", "namespace": "zenmdm", "cluster": "mdm-foundationdb-1621887305593585", "error": "Get \"https://10.254.22.105:8080/substitutions\": x509: cannot validate certificate for 10.254.22.105 because it doesn't contain any IP SANs"}
github.com/go-logr/zapr.(*zapLogger).Error
	/go/pkg/mod/github.com/go-logr/zapr@v0.1.0/zapr.go:128
github.com/FoundationDB/fdb-kubernetes-operator/controllers.(*FoundationDBClusterReconciler).Reconcile
	/workspace/controllers/cluster_controller.go:152
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.4.0/pkg/internal/controller/controller.go:256
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.4.0/pkg/internal/controller/controller.go:232
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.4.0/pkg/internal/controller/controller.go:211
k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1
	/go/pkg/mod/k8s.io/apimachinery@v0.17.0/pkg/util/wait/wait.go:152
k8s.io/apimachinery/pkg/util/wait.JitterUntil
	/go/pkg/mod/k8s.io/apimachinery@v0.17.0/pkg/util/wait/wait.go:153
k8s.io/apimachinery/pkg/util/wait.Until
	/go/pkg/mod/k8s.io/apimachinery@v0.17.0/pkg/util/wait/wait.go:88
```

**Comment by johscheuer** [comment]
You also want to set in that case to tell the operator not to check the server certificate:

```yaml
        - name: DISABLE_SIDECAR_TLS_CHECK         
          value: "1"
```

We update our documentation to make that more clear.

**Comment by tangerine-tt** [comment]
Hi Johscheuer, I tried your suggestion but with a higher version of fdb operator (0.31.1) and fdb image (6.2.29).   I can confirm that the fdb-certs are created in /tmp in the main image.  Also, checked env. variables for operator, sidecar and main image and they are all there.  But I am still getting the same empty config map issues as I see the sidecar is stuck in this log:
2021-05-25 21:15:25,335 INFO Detected change to certificates
2021-05-25 21:15:35,343 INFO Reloading certificates
Basically, it keeps looping and never get out of it.
What could be the issue?  I used the example secret that's in the documentation of the operator.

**Comment by johscheuer** [comment]
An empty ConfigMap is probably an issue with the operator (or rather with the config) could you take a look if there are any log lines in the operator output that indicates an error (e.g. has the log level `error` sometimes). The reloading loop shouldn't be an issue that is only confusing (and will bee fixed in the next `6.3.12` release). Depending on what revision of the repository you are using the certificates might not valid anymore. You can either use the cert from the master branch or just use these script from the repository: https://github.com/FoundationDB/fdb-kubernetes-operator/tree/master/config/test-certs

**Comment by tangerine-tt** [comment]
Here is the log from the operator that has error:
{"level":"error","ts":1622039694.6145732,"logger":"controller","msg":"Error in reconciliation","subReconciler":"controllers.GenerateInitialClusterFile","namespace":"test-fdborg","cluster":"sample-cluster","error":"open /tmp/fdb-certs/tls.crt: no such file or directory","stacktrace":"github.com/go-logr/zapr.(*zapLogger).Error\n\t/go/pkg/mod/github.com/go-logr/zapr@v0.2.0/zapr.go:132\nsigs.k8s.io/controller-runtime/pkg/log.(*DelegatingLogger).Error\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.8.3/pkg/log/deleg.go:144\ngithub.com/FoundationDB/fdb-kubernetes-operator/controllers.(*FoundationDBClusterReconciler).Reconcile\n\t/workspace/controllers/cluster_controller.go:171\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.8.3/pkg/internal/controller/controller.go:298\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.8.3/pkg/internal/controller/controller.go:253\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1.2\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.8.3/pkg/internal/controller/controller.go:216\nk8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext.func1\n\t/go/pkg/mod/k8s.io/apimachinery@v0.20.4/pkg/util/wait/wait.go:185\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1\n\t/go/pkg/mod/k8s.io/apimachinery@v0.20.4/pkg/util/wait/wait.go:155\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil\n\t/go/pkg/mod/k8s.io/apimachinery@v0.20.4/pkg/util/wait/wait.go:156\nk8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/go/pkg/mod/k8s.io/apimachinery@v0.20.4/pkg/util/wait/wait.go:133\nk8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext\n\t/go/pkg/mod/k8s.io/apimachinery@v0.20.4/pkg/util/wait/wait.go:185\nk8s.io/apimachinery/pkg/util/wait.UntilWithContext\n\t/go/pkg/mod/k8s.io/apimachinery@v0.20.4/pkg/util/wait/wait.go:99"}
{"level":"error","ts":1622039694.6153457,"logger":"controller-runtime.manager.controller.foundationdbcluster","msg":"Reconciler error","reconciler group":"apps.foundationdb.org","reconciler kind":"FoundationDBCluster","name":"sample-cluster","namespace":"test-fdborg","error":"open /tmp/fdb-certs/tls.crt: no such file or directory","stacktrace":"github.com/go-logr/zapr.(*zapLogger).Error\n\t/go/pkg/mod/github.com/go-logr/zapr@v0.2.0/zapr.go:132\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.8.3/pkg/internal/controller/controller.go:302\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.8.3/pkg/internal/controller/controller.go:253\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1.2\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.8.3/pkg/internal/controller/controller.go:216\nk8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext.func1\n\t/go/pkg/mod/k8s.io/apimachinery@v0.20.4/pkg/util/wait/wait.go:185\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1\n\t/go/pkg/mod/k8s.io/apimachinery@v0.20.4/pkg/util/wait/wait.go:155\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil\n\t/go/pkg/mod/k8s.io/apimachinery@v0.20.4/pkg/util/wait/wait.go:156\nk8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/go/pkg/mod/k8s.io/apimachinery@v0.20.4/pkg/util/wait/wait.go:133\nk8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext\n\t/go/pkg/mod/k8s.io/apimachinery@v0.20.4/pkg/util/wait/wait.go:185\nk8s.io/apimachinery/pkg/util/wait.UntilWithContext\n\t/go/pkg/mod/k8s.io/apimachinery@v0.20.4/pkg/util/wait/wait.go:99"}
{"level":"info","ts":1622039700.350755,"logger":"controller","msg":"Generating initial cluster file","namespace":"test-fdborg","cluster":"sample-cluster"}
{"level":"info","ts":1622039700.3511553,"logger":"controller","msg":"Reconciliation terminated early","namespace":"test-fdborg","name":"sample-cluster","lastAction":"controllers.GenerateInitialClusterFile"}
{"level":"error","ts":1622039700.3512,"logger":"controller","msg":"Error in reconciliation","subReconciler":"controllers.GenerateInitialClusterFile","namespace":"test-fdborg","cluster":"sample-cluster","error":"open /tmp/fdb-certs/tls.crt: no such file or directory","stacktrace":"github.com/go-logr/zapr.(*zapLogger).Error\n\t/go/pkg/mod/github.com/go-logr/zapr@v0.2.0/zapr.go:132\nsigs.k8s.io/controller-runtime/pkg/log.(*DelegatingLogger).Error\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.8.3/pkg/log/deleg.go:144\ngithub.com/FoundationDB/fdb-kubernetes-operator/controllers.(*FoundationDBClusterReconciler).Reconcile\n\t/workspace/controllers/cluster_controller.go:171\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.8.3/pkg/internal/controller/controller.go:298\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.8.3/pkg/internal/controller/controller.go:253\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1.2\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.8.3/pkg/internal/controller/controller.go:216\nk8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext.func1\n\t/go/pkg/mod/k8s.io/apimachinery@v0.20.4/pkg/util/wait/wait.go:185\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1\n\t/go/pkg/mod/k8s.io/apimachinery@v0.20.4/pkg/util/wait/wait.go:155\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil\n\t/go/pkg/mod/k8s.io/apimachinery@v0.20.4/pkg/util/wait/wait.go:156\nk8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/go/pkg/mod/k8s.io/apimachinery@v0.20.4/pkg/util/wait/wait.go:133\nk8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext\n\t/go/pkg/mod/k8s.io/apimachinery@v0.20.4/pkg/util/wait/wait.go:185\nk8s.io/apimachinery/pkg/util/wait.UntilWithContext\n\t/go/pkg/mod/k8s.io/apimachinery@v0.20.4/pkg/util/wait/wait.go:99"}
{"level":"error","ts":1622039700.3513758,"logger":"controller-runtime.manager.controller.foundationdbcluster","msg":"Reconciler error","reconciler group":"apps.foundationdb.org","reconciler kind":"FoundationDBCluster","name":"sample-cluster","namespace":"test-fdborg","error":"open /tmp/fdb-certs/tls.crt: no such file or directory","stacktrace":"github.com/go-logr/zapr.(*zapLogger).Error\n\t/go/pkg/mod/github.com/go-logr/zapr@v0.2.0/zapr.go:132\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.8.3/pkg/internal/controller/controller.go:302\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.8.3/pkg/internal/controller/controller.go:253\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1.2\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.8.3/pkg/internal/controller/controller.go:216\nk8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext.func1\n\t/go/pkg/mod/k8s.io/apimachinery@v0.20.4/pkg/util/wait/wait.go:185\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1\n\t/go/pkg/mod/k8s.io/apimachinery@v0.20.4/pkg/util/wait/wait.go:155\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil\n\t/go/pkg/mod/k8s.io/apimachinery@v0.20.4/pkg/util/wait/wait.go:156\nk8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/go/pkg/mod/k8s.io/apimachinery@v0.20.4/pkg/util/wait/wait.go:133\nk8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext\n\t/go/pkg/mod/k8s.io/apimachinery@v0.20.4/pkg/util/wait/wait.go:185\nk8s.io/apimachinery/pkg/util/wait.UntilWithContext\n\t/go/pkg/mod/k8s.io/apimachinery@v0.20.4/pkg/util/wait/wait.go:99"}

**Comment by tangerine-tt** [comment]
seems like the tls.crt is not mounted.

**Comment by johscheuer** [comment]
Could you try to mount it and then try it again?

**Comment by tangerine-tt** [comment]
Hi Joscheuer, I managed to get the feb-certs mounted in the operator pod and also have the env. variables setup.  In fact, those env. variables have been setup on the fdb and the sidecar as well as the fdb-certs are mounted in /tmp on all pods.  The DB is up and I have set the following in the cluster deployment yaml.
mainContainer:
   enableTls: true
sidecarContainer:
   enableTls: true

The db is up and running, in side the storage pod, I can run fdbcli and status show db is available.   However, looking at the configMap, I don't see any TLS in the cluster configuration.  Is there any way to check whether the DB is running in TLS mode other then trying to run a fdbcli from outside of the cluster to try to connect to it with TLS?

**Comment by tangerine-tt** [comment]
Also, I get a bounceProcess error in the operator log, but the DB is up and running.  FDBcli status detail show db is available.  Just no TLS and the bouncProcess error in the operator log.  So, is it because it failed to bounce so the TLS is not configured?

**Comment by tangerine-tt** [comment]
Generated from foundationdbcluster-controller
177 times in the last 10 minutes
(combined from similar events): Spec require a bounce of some processes, but the cluster has only been up for 340.007000 seconds
How long it takes to make it bounce?

**Comment by johscheuer** [comment]
> The db is up and running, in side the storage pod, I can run fdbcli and status show db is available. However, looking at the configMap, I don't see any TLS in the cluster configuration. Is there any way to check whether the DB is running in TLS mode other then trying to run a fdbcli from outside of the cluster to try to connect to it with TLS?

You should see in the connection string a `:tls` suffix for each address: 

```bash
 $kubectl get cm sample-cluster-config -o jsonpath='{.data.cluster\-file}'
sample_cluster:AB4NE7K2t92ZPSzXsLWPFGf9YygskP3t@10.1.12.30:4500:tls,10.1.12.31:4500:tls,10.1.12.33:4500:tls
```

You can also check the current `fdbserver` process with `ps` or check the `/var/dynamic-conf/fdbmonitor.conf` file to see the pls config:

```bash
$ cat /var/dynamic-conf/fdbmonitor.conf
[general]
kill_on_configuration_change = false
restart_delay = 60
[fdbserver.1]
command = /usr/bin/fdbserver
cluster_file = /var/fdb/data/fdb.cluster
seed_cluster_file = /var/dynamic-conf/fdb.cluster
public_address = 10.1.12.33:4500:tls # <--- see the :tls
class = storage
logdir = /var/log/fdb-trace-logs
loggroup = sample-cluster
datadir = /var/fdb/data
locality_instance_id = storage-3
locality_machineid = sample-cluster-storage-3
locality_zoneid = sample-cluster-storage-3
knob_disable_posix_kernel_aio=1
```

The `Spec require a bounce of some processes` indicates that the operator needs to restart some processes but the cluster is not long enough up to consider a kill as save (the default value is `500s` you can change that time by setting `minimumUptimeSecondsForBounce` in the cluster spec). I guess that's a bug how we initialise the cluster it seems like the cluster listens on both addresses and then the operator needs to restart all processes in a way that they only listen on TLS. That should be resolved after `500s`. I open an issue for that since I assume that's not the expected behaviour.

**Comment by tangerine-tt** [comment]
ok, I get it to work now.
I found that if I use an expired certificate (example, the one on the website), for the operator version that I tried (0.31.1), it will cause the operator to destabilized (nil pointer exception, and crash).  Using a higher version of operator (0.35.1) will print out a meaningful message saying the certificate has expired and will not crash.  Using a correct certificate will just work without issues.

**Comment by johscheuer** [comment]
Great 👍 Sorry for the rather rough experience with setting up TLS. We try our best to document all these steps (or if you want to contribute these steps I'm very happy to look at the PR). If there is nothing left we could close this issue, I hope I open follow up issues for the more specific tasks but feel free to open other issues if you encounter anything else.

**Comment by tangerine-tt** [comment]
Thanks, yes this particular one can be closed.
